packages-exp/auth-exp/src/core/util/popup.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getUA } from '@firebase/util';
+
+import { AuthErrorCode } from '../errors';
+import { assert } from './assert';
+
+const BASE_POPUP_OPTIONS = {
+  location: 'yes',
+  resizable: 'yes',
+  statusbar: 'yes',
+  toolbar: 'no'
+};
+
+const DEFAULT_WIDTH = 500;
+const DEFAULT_HEIGHT = 600;
+
+const CHROME_IOS_UA = 'crios/';
+const FIREFOX_UA = 'firefox/';
+const FIREFOX_EMPTY_URL = 'http://localhost';
+
+export class AuthPopup {
+  associatedEvent: string | null = null;
+
+  constructor(readonly window: Window) {}
+
+  close(): void {
+    try {
+      this.window.close();
+    } catch (e) {}
+  }
+}
+
+/** Wrapper so that we can stub the UA in tests */
+export const _uaGetter = {
+  getUA
+};
+
+export function _open(
+  appName: string,
+  url?: string,
+  name?: string,
+  width = DEFAULT_WIDTH,
+  height = DEFAULT_HEIGHT
+): AuthPopup {
+  const top = Math.min((window.screen.availHeight - height) / 2, 0).toString();
+  const left = Math.min((window.screen.availWidth - width) / 2, 0).toString();
+  let target = '';
+
+  const options: { [key: string]: string } = {
+    ...BASE_POPUP_OPTIONS,
+    width: width.toString(),
+    height: height.toString(),
+    top,
+    left
+  };
+
+  // Chrome iOS 7 and 8 is returning an undefined popup win when target is
+  // specified, even though the popup is not necessarily blocked.
+  const ua = _uaGetter.getUA().toLowerCase();
```

Comment:
use the helpers in firebase/util please

Reply:
This is just so that we can stub this out in test without importing all of firebase/util.

The `import * as <something> from '...'` works for stubbing BUT it also will break tree shaking (the entire module will be imported).

This is fine when there's only one or two functions in that module, but for something like firebase/util, it's a no-go

Reply:
we only need to do the import * as something in the tests, righty? in the implementation you can just pull in the method you want, either way we pull the UA helpers elsewhere